Add example howto repartion a VM disk at boot time (#136)
syseleven/grav-docs-syseleven-stack: user/pages/03.Howtos/15.repartition-vm-disk/docs.en.md

This Document will show you the essential steps to repartition your VM disk. For this purpose we are using an Ubuntu 18.04 server as example with the `m1c.tiny` [flavor](../../04.Reference/03.compute/docs.en.md).

We strongly recommend to use the automatic approaches but for instructional use we are also covering the manual process here.

### Prerequisites

* You need to have the login data for the SysEleven Stack API (user name and passphrase)
* Knowledge how to utilise a terminal/SSH and SSH-keys.

### Examples

By default [cloudinit](https://cloudinit.readthedocs.io/en/latest/) will grow your VM image partition size to the complete size of the ephemeral disk (which is defined by the flavor you used for your VM). To avoid this behaviour we can tell cloudinit to behave differently.

Using the OpenStack dashboard to start your VM you may provide the cloudinit snippets from the following sections in the `Customization Script` box located in the `Configuration` tab. If you prefer to use the CLI to bring up the VM, you can use the `--user-data` option to provide the cloud-config file containing the snippet.

> The snippets shown in the examples were written and tested for an Ubuntu 18.04 VM with a 50GiB ephemeral disk. Be aware when working with instance or volume snapshots that you should take additional care with your already pre-occupied disk space.

Reviewer: ```suggestion
The snippets shown in the examples were written and tested for an Ubuntu 18.04 VM with a 50 GiB ephemeral disk. Be aware when working with instance or volume snapshots that you should take additional care with your already pre-occupied disk space.
```

tmpfs           200M     0  200M   0% /run/user/1000
```

> ### Example 2 : (Manually) configure partitions

Reviewer: ```suggestion
### Example 3 : (Manually) configure partitions
```
I'm all for sequential counting. Apart from that, I see the didactical order inversed. Shouldn't we start with the simplest approach and go towards the more sophisticated (but nonetheless recommended) ones?

- "for i in $(cat /partition_nums.txt); do mount -t ext4 /dev/vda${i} /mnt/disks/data${i}; done"
- "for i in $(cat /partition_nums.txt); do echo LABEL=DATA${i} /mnt/disks/data${i} ext4 defaults 0 0 | sudo tee -a /etc/fstab; done"
# System is not aware of new space on main partition. Needs refresh.
- "resize2fs /dev/vda1"
```

It will resize the main partition to 10 GiB and create 2 new partitions (10 GiB and 30 GiB), create an ext-4 filesystem and mount them.

After provisioning we may for our new created partitions inside of the VM :

```shell
ubuntu@partition-test:~$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            985M     0  985M   0% /dev
tmpfs           200M  648K  199M   1% /run
/dev/vda1       8.9G  1.1G  7.9G  12% /
tmpfs           997M     0  997M   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           997M     0  997M   0% /sys/fs/cgroup
/dev/vda15      105M  3.6M  101M   4% /boot/efi
/dev/vda2       9.2G   37M  8.6G   1% /mnt/disks/data2
/dev/vda3        28G   45M   26G   1% /mnt/disks/data3
tmpfs           200M     0  200M   0% /run/user/1000
```

#### Example 3 : (Automatically) configure partitions at boot

> Following snippet will limit the growing of the ephemeral disk at 10GB and use the spare space for creating another partition :

Reviewer: ```suggestion
Following snippet will limit the growing of the ephemeral disk at 10GiB and use the spare space for creating another partition :
```

> Following snippet will limit the growing of the ephemeral disk at 10GiB and use the spare space for creating another partition :

Reviewer: ```suggestion
Following snippet will limit the growing of the ephemeral disk at 10 GiB and use the spare space for creating another partition :
```

8303 Linux x86 root (/)    8304 Linux x86-64 root (/  8305 Linux ARM64 root (/)
8306 Linux /srv            8307 Linux ARM32 root (/)  8400 Intel Rapid Start
8e00 Linux LVM             a000 Android bootloader    a001 Android bootloader 2
a002 Android boot          a003 Android recovery      a004 Android misc
a005 Android metadata      a006 Android system        a007 Android cache
a008 Android data          a009 Android persistent    a00a Android factory
a00b Android fastboot/ter  a00c Android OEM           a500 FreeBSD disklabel
a501 FreeBSD boot          a502 FreeBSD swap          a503 FreeBSD UFS
a504 FreeBSD ZFS           a505 FreeBSD Vinum/RAID    a580 Midnight BSD data
a581 Midnight BSD boot     a582 Midnight BSD swap     a583 Midnight BSD UFS
a584 Midnight BSD ZFS      a585 Midnight BSD Vinum    a600 OpenBSD disklabel
a800 Apple UFS             a901 NetBSD swap           a902 NetBSD FFS
a903 NetBSD LFS            a904 NetBSD concatenated   a905 NetBSD encrypted
a906 NetBSD RAID           ab00 Recovery HD           af00 Apple HFS/HFS+
af01 Apple RAID            af02 Apple RAID offline    af03 Apple label
```

> After provisioning we may for our new created 40 GB partition inside of the VM :

Reviewer: ```suggestion
After provisioning we may for our new created 40 GiB partition inside of the VM :
```

a504 FreeBSD ZFS           a505 FreeBSD Vinum/RAID    a580 Midnight BSD data
a581 Midnight BSD boot     a582 Midnight BSD swap     a583 Midnight BSD UFS
a584 Midnight BSD ZFS      a585 Midnight BSD Vinum    a600 OpenBSD disklabel
a800 Apple UFS             a901 NetBSD swap           a902 NetBSD FFS
a903 NetBSD LFS            a904 NetBSD concatenated   a905 NetBSD encrypted
a906 NetBSD RAID           ab00 Recovery HD           af00 Apple HFS/HFS+
af01 Apple RAID            af02 Apple RAID offline    af03 Apple label
```

After provisioning we may for our new created 40 GB partition inside of the VM :

```shell
ubuntu@partition-test:~$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            984M     0  984M   0% /dev
tmpfs           200M  644K  199M   1% /run
/dev/vda1       9.6G  1.1G  8.5G  12% /
tmpfs           997M     0  997M   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           997M     0  997M   0% /sys/fs/cgroup
/dev/vda15      105M  6.6M   98M   7% /boot/efi
/dev/vda2        40G   49M   38G   1% /mnt/disks/data
tmpfs           200M     0  200M   0% /run/user/1000
```

### Conclusion

> We have started a VM using custom user data which prevents the growing of the VM root partition to the complete size of the ephemeral disk. The resulting free disk space was used to setup custom partitions on which we created an ext4 filesystem, mounted it and made our changes persistent for reboots.

Reviewer: Aren't partition and filesystem changes always "persistent for reboots"? Speaking of reboots, I'm not sure what happens in variant #2 when the system gets rebooted multiple times and the space is already used.

Author: I think the mount is not persistent, I will give it some more testing and revisit the text afterwards

Author: Cloudinit is just executed once at VM creation. Thus I do not see this being a problem. Only if you refer to instance / volume snapshots I get the point. Will add a warning about that. Partition and Filesystem changes are persistent yes, but mount is not, updating the text accordingly =)
